### ISSUE TITLE
Buildah is injecting /bin/bash into cmd and should not be

### ISSUE THREAD

**Issue by TomSweeneyRedHat** [projectatomic/buildah#600]
**Description**

As reported by @pixdrift in #593 
@rhatdan @TomSweeneyRedHat this isn't a complete fix. When using buildah native it is injecting a spurious /bin/bash into cmd which shouldn't be there.

What's occuring then is that the cmd is being appended to the entrypoint. Changing entrypoint to /bin/sh -c will match docker entrypoint behaviour, but doesn't fix buildah always injecting /bin/bash into cmd.

**Steps to reproduce the issue:**
build and image using buildah bud or buildah cmd.  Then run 'buildah inspect' on the image and note the extraneous /bin/bash in the Cmd.

**Describe the results you received:**
See description.

**Describe the results you expected:**

No extraneous /bin/bash.  This can can cause the container to not be  run appropriately.

**Output of `buildah version`:**

Buildah v0.16

**Comment by pixdrift** [comment]
Thanks @TomSweeneyRedHat, I had the ticket half written up.

**Comment by TomSweeneyRedHat** [comment]
@pixdrift, that's unusual, I usually don't have fast fingers in the morning before my tea!  ;^)   I tried to assign this one to you, but GitHub for some reason wouldn't let me.

**Comment by pixdrift** [comment]
@TomSweeneyRedHat, I tried to track this down last night and had trouble finding where the value was coming from. Keen to see what you (or whoever provides the fix) comes up with so I can better understand the components involved in the earlier part of the creation process. I ended up on `config.go` last night when looking.

**Comment by rhatdan** [comment]
Is the /bin/bash defined in the container image you are building from?

**Comment by pixdrift** [comment]
@rhatdan, sadly yes :blush:

This is where it was coming from. The problem in #593 was that it was coming from the parent container (centos 7) and getting appended to the entrypoint added.

I guess blanking values from the parent container is an exercise for the user. Just checking the bud behaviour to see if it matches.

**Comment by pixdrift** [comment]
@rhatdan, give me 5 mins.. I will try and get a more complete picture of behaviour with/without options

**Comment by pixdrift** [comment]
Source container (centos:7):
```
# buildah inspect centos:7 | jq '.Docker.config.Cmd'
[
  "/bin/bash"
]
# buildah inspect centos:7 | jq '.OCIv1.config.Cmd'
[
  "/bin/bash"
]
```

**Test 1:** buildah bud the following (no entrypoint):
```
FROM centos:7
WORKDIR /tmp
```

Result:
# buildah inspect 215c | jq '.Docker.config.Cmd'
[
  "/bin/bash"
]
# buildah inspect 215c | jq '.OCIv1.config.Cmd'
[
  "/bin/bash"
]
```

**Test 2:** buildah bud the following (entrypoint):
```
FROM centos:7
ENTRYPOINT /bin/bash
WORKDIR /tmp
```

Result:
```
# buildah inspect 87d5 | jq '.OCIv1.config.Cmd'
null
# buildah inspect 87d5 | jq '.Docker.config.Cmd'
null
```

**Test 3:** buildah native the following (no entrypoint):
```
#!/bin/bash
ctr=$(buildah from centos:7)
buildah commit $ctr pix_noentry
```

Result:
```
# buildah inspect pix_noentry | jq '.OCIv1.config.Cmd'
[
  "/bin/bash"
]
# buildah inspect pix_noentry | jq '.Docker.config.Cmd'
[
  "/bin/bash"
]
```

**Test 4:** buildah native the following (entrypoint):
```
# buildah inspect pix_entry | jq '.Docker.config.Cmd'
[
  "/bin/bash"
]
# buildah inspect pix_entry | jq '.OCIv1.config.Cmd'
[
  "/bin/bash"
]

```

In summary, Test 2 and Test 4 should produce the same result.

**Comment by rhatdan** [comment]
Could you also do the two tests with Docker.

**Comment by rhatdan** [comment]
It looks like specifying ENTRYPOINT has a side effect of clearing command.  If docker does the same, then we should do it.

**Comment by pixdrift** [comment]
@rhatdan ack and agree. Will fire up Docker now.

**Comment by pixdrift** [comment]
**Test 5:** docker build the following (no entrypoint)
```
FROM centos:7
WORKDIR /tmp
```

Result:
```
# docker inspect b33 | jq '.[].Config.Cmd'
[
  "/bin/bash"
]
```

**Test 6:** docker build the following (entrypoint)
```
FROM centos:7
ENTRYPOINT /bin/bash
WORKDIR /tmp
```

Result:
```
# docker inspect 93e | jq '.[].Config.Cmd'
null
```

@rhatdan, results. Test 4 doesn't align.

**Comment by rhatdan** [comment]
Thanks @pixdrift 
Looks like setting ENTRYPOINT should nil out CMD.  

buildah config --entrypoint foo

Should result in cmd == Null

buildah config --entrypoint foo --command bar
Should result in cmd == bar

buildah config --workingdir /tmp

Should result in cmd == Parent Container CMD.

**Comment by pixdrift** [comment]
PR will be ~24 hours away because i'm running close to midnight here and need to wrap up. Happy for someone else to put the PR together during your daylight hours today, if not, I will pick it up tomorrow my time. Cheers.

**Comment by rhatdan** [comment]
I opened a PR for it.  Thanks @pixdrift

**Comment by TomSweeneyRedHat** [comment]
Just to be sure, if someone specifies both the --cmd and --entry-point parameters, both should be retained.  In that case the Cmd should not be set to nil, but to what ever the user passed in.   Holldr if that's not right.

**Comment by rhatdan** [comment]
Yes

**Comment by pixdrift** [comment]
Just thought of another potential test/edge case
`buildah --entrypoint "" $ctr`

You're setting the entrypoint.. in this case to an empty value, should that also blank/clear cmd?

**Comment by rhatdan** [comment]
If you do that in a Dockerfile and run `docker build` on it, what happens?

**Comment by pixdrift** [comment]
@rhatdan I will do some more tests including Docker

**Comment by pixdrift** [comment]
Test 7: docker build the following (empty entrypoint)
```
FROM centos:7
ENTRYPOINT ""
WORKDIR /tmp

```

Result:
```
# docker inspect a50 | jq '.[].Config.Cmd'
null
```

**Comment by adelton** [comment]
I've just hit the issue attempting to test FreeIPA container build. The `ENTRYPOINT` value is injected to `RUN` commands as well, something docker does not do in build time:

```
# cat Dockerfile
FROM registry.fedoraproject.org/fedora:26
RUN uuidgen > /build-id-pre-entrypoint
ENTRYPOINT [ "/non/existent/entrypoint" ]
RUN uuidgen > /build-id-post-entrypoint
```

```
# docker build -t test-entrypoint .
Sending build context to Docker daemon 2.048 kB
Step 1/4 : FROM registry.fedoraproject.org/fedora:26
 ---> e569af8fd8d7
Step 2/4 : RUN uuidgen > /build-id-pre-entrypoint
 ---> Using cache
 ---> cbe765590c1d
Step 3/4 : ENTRYPOINT /non/existent/entrypoint
 ---> Using cache
 ---> 61e435f6b309
Step 4/4 : RUN uuidgen > /build-id-post-entrypoint
 ---> Using cache
 ---> 5954df23046e
Successfully built 5954df23046e
# docker run --entrypoint cat test-entrypoint /build-id-post-entrypoint
c98ac630-132e-4706-89db-4e944ee727c3
```

```
# buildah bud -t test-entrypoint .
STEP 1: FROM registry.fedoraproject.org/fedora:26
STEP 2: RUN uuidgen > /build-id-pre-entrypoint
STEP 3: ENTRYPOINT [ "/non/existent/entrypoint" ]
STEP 4: RUN uuidgen > /build-id-post-entrypoint
container_linux.go:348: starting container process caused "exec: \"/non/existent/entrypoint\": stat /non/existent/entrypoint: no such file or directory"
error building at step {Env:[DISTTAG=f26container FGC=f26  PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin] Command:run Args:[uuidgen > /build-id-post-entrypoint] Flags:[] Attrs:map[] Message:RUN uuidgen > /build-id-post-entrypoint Original:RUN uuidgen > /build-id-post-entrypoint}: exit status 1
```

In FreeIPA case we don't use nonexistent path for `ENTRYPOINT` but we specify script which ecpects only a limited set of parameters, so subsequent `RUN`s fail because they are not in the list.

Is it expected that `ENTRYPOINT` would need to be the last command in the `Dockerfile`?

**Comment by TomSweeneyRedHat** [comment]
@adelton The ENTRYPOINT command shouldn't need to be the last command/line in the Dockerfile.  If you find otherwise, please let us know.

**Comment by pixdrift** [comment]
I think this is related but is a separate issue.

Buildah bud `RUN` should be treated differently to command line `buildah run`.

**Comment by adelton** [comment]
@pixdrift, thanks for reproducing it. Should I file a separate issue or is it OK to track it here?

**Comment by pixdrift** [comment]
@adelton i'll raise a separate issue for it now as this issue was just closed out by a PR.

**Comment by TomSweeneyRedHat** [comment]
@adelton and @pixdrift , yes please a separate issue and ty!

**Comment by pixdrift** [comment]
Opened issue #607 

Should be good to close this one out as PR #602 resolved.

**Comment by rhatdan** [comment]
I don't agree buildah run and buildah bud Dockerfile RUN, should work the same.  The basic idea of the buildah run command is to be an equivalent BASH version of the Dockerfile RUN command.  @pixdrift Why do you think they should behave differently?  

BTW I agree that this is a bug, we should match the Docker behaviour.

**Comment by pixdrift** [comment]
@rhatdan ? I don't think it, it's just from what I have determined from testing.

**Comment by pixdrift** [comment]
@TomSweeneyRedHat, should be good to close this, was resolved with PR #602 

@rhatdan, agree re: RUN behaviour, see the comment you were responding to.

To clarify, the following _don't_ use entrypoint value and behave the same
```
docker build RUN
buildah bud RUN
buildah run
```

The following _do_ use entrypoint value and behave the same
```
docker run
podman run
```

**Comment by TomSweeneyRedHat** [comment]
Closing based on #602.  Thanks to all for the discussion, especially @pixdrift for starting it up.
